contrib/babelfishpg_tsql/src/procedures.c

```diff
@@ -1283,145 +1287,299 @@ sp_describe_undeclared_parameters_internal(PG_FUNCTION_ARGS)
-		"WHEN T2.name COLLATE sys.database_default = \'ntext\' THEN 2147483646 "
-		"ELSE CAST( C.max_length AS INT ) "
+		", CASE T2.name COLLATE sys.database_default "
+		"WHEN \'nvarchar\' THEN "
+			"CASE WHEN CASE "
+				"WHEN a.atttypmod != -1 THEN "
+					"sys.tsql_type_max_length_helper(type, a.attlen, a.atttypmod) "
+				"ELSE "
+					"sys.tsql_type_max_length_helper(type, a.attlen, t.typtypmod) "
+			"END  = -1 THEN 65535 "
+			"ELSE CASE "
+				"WHEN a.atttypmod != -1 THEN "
+					"sys.tsql_type_max_length_helper(type, a.attlen, a.atttypmod) "
+				"ELSE "
+					"sys.tsql_type_max_length_helper(type, a.attlen, t.typtypmod) "
+			"END  "
+			"END "
+		"WHEN \'varbinary\' THEN "
+			"CASE WHEN CASE "
+				"WHEN a.atttypmod != -1 THEN "
+					"sys.tsql_type_max_length_helper(type, a.attlen, a.atttypmod) "
+				"ELSE "
+					"sys.tsql_type_max_length_helper(type, a.attlen, t.typtypmod) "
+			"END  = -1 THEN 65535 "
+			"ELSE CASE "
+				"WHEN a.atttypmod != -1 THEN "
+					"sys.tsql_type_max_length_helper(type, a.attlen, a.atttypmod) "
+				"ELSE "
+					"sys.tsql_type_max_length_helper(type, a.attlen, t.typtypmod) "
+			"END  "
+			"END "
+		"WHEN \'varchar\' THEN "
+			"CASE WHEN CASE "
+				"WHEN a.atttypmod != -1 THEN "
+					"sys.tsql_type_max_length_helper(type, a.attlen, a.atttypmod) "
+				"ELSE "
+					"sys.tsql_type_max_length_helper(type, a.attlen, t.typtypmod) "
+			"END  = -1 THEN 65535 "
+			"ELSE CASE "
+				"WHEN a.atttypmod != -1 THEN "
+					"sys.tsql_type_max_length_helper(type, a.attlen, a.atttypmod) "
+				"ELSE "
+					"sys.tsql_type_max_length_helper(type, a.attlen, t.typtypmod) "
+			"END  "
+			"END "
+		"WHEN \'decimal\' THEN 17 "
+		"WHEN \'numeric\' THEN 17 "
+		"WHEN \'xml\' THEN 8100 "
+		"WHEN \'image\' THEN 2147483647 "
+		"WHEN \'text\' THEN 2147483647 "
+		"WHEN \'ntext\' THEN 2147483646 "
+		"ELSE CAST( CASE "
+				"WHEN a.atttypmod != -1 THEN "
+					"sys.tsql_type_max_length_helper(type, a.attlen, a.atttypmod) "
+				"ELSE "
+					"sys.tsql_type_max_length_helper(type, a.attlen, t.typtypmod) "
+			"END  AS INT ) "
 		"END "					/* AS "suggested_tds_length" */
-		"FROM sys.objects O, sys.columns C, sys.types T, sys.types T2 "
-		"WHERE O.object_id = C.object_id "
-		"AND C.user_type_id = T.user_type_id "
-		"AND C.name = \'%s\' COLLATE sys.database_default " /* -- INPUT column name */
-		"AND T.system_type_id = T2.user_type_id "	/* -- To get system dt
-													 * name. */
-		"AND O.name = \'%s\' COLLATE sys.database_default " /* -- INPUT table name */
-		"AND O.schema_id = %d " /* -- INPUT schema Oid */
-		"AND O.type = \'U\'";	/* -- User tables only for the time being */
-
-		char	   *query = psprintf(tempq,
-									 undeclaredparams->targetcolnames[undeclaredparams->paramindexes[call_cntr]],
-									 undeclaredparams->tablename,
-									 undeclaredparams->schemaoid);
+		"FROM sys.types AS T2 "
+		"JOIN pg_attribute AS a ON a.atttypid = T2.user_type_id "
```

Comment:
pg_attribute join sys.type would make more sense

Reply:
Updated in latest rev.